### PR TITLE
KEP-508 secure the whitelist functionality

### PR DESCRIPTION
**Pull request by rnistuk**
Black list and verifying uuid's against signatures created with Bluzelle owned private keys.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -149,7 +149,8 @@ The configuration file is a JSON format file, as seen in the following example:
         "logfile_dir" : "logs/",
         "logfile_rotation_size" : "64K",
         "logfile_max_size" : "640K",
-        "debug_logging" : false
+        "debug_logging" : false,
+        "security_enabled" : false
```
Comment (ebruck) [inline README.md]: wouldn't a name like "peer_validation" be better?

File: bootstrap/bootstrap_peers.cpp [hunk 2]
```diff
@@ -103,6 +105,18 @@ bootstrap_peers::initialize_peer_list(const Json::Value& root, bzn::peers_list_t
             continue;
         }
 
+        if (this->is_security_enabled())
+        {
+            // At this point we cannot validate uuids as we do not have
+            // signatures for all of them, so we simply ignore blacklisted
+            // uuids
+            if(bzn::utils::blacklist::is_blacklisted(uuid))
```
Comment (ebruck) [inline bootstrap/bootstrap_peers.cpp]: space after if

File: raft/raft.cpp [hunk 3]
```diff
@@ -423,6 +424,33 @@ raft::send_session_error_message(std::shared_ptr<bzn::session_base> session, con
...
+    // uuid's are signed as text files, so we must append the end of line character
+    const auto uuid{peer["uuid"].asString() + "\x0a"};
+    if (!peer.isMember("signature")  || peer["signature"].asString().empty())
+    {
+        this->send_session_error_message(session, ERROR_INVALID_SIGNATURE);
+        return false;
+    }
+
+    const auto signature{peer["signature"].asString()};
+    if(!bzn::utils::crypto::verify_signature(bzn::utils::crypto::retrieve_bluzelle_public_key_from_contract(), signature, uuid))
```
Comment (ebruck) [inline raft/raft.cpp]: space after if

File: utils/crypto.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+     * This function is used by base_64_decode to determine the amount of memory required
+     * for the binary buffer represented by the base64 encoded string.
+     *
+     * @param base_64_input base64 encoded string
+     * @return a size_t containing the number of bytes expected in the decoded string.
+    */
+    size_t
+    calculate_decode_length(const std::string& base_64_input)
+    {
+        const auto len{base_64_input.size()};
```
Comment (ebruck) [inline utils/crypto.cpp]: change auto to size_t

File: utils/crypto.cpp [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    /**
+     * This function is used by base_64_decode to determine the amount of memory required
+     * for the binary buffer represented by the base64 encoded string.
+     *
+     * @param base_64_input base64 encoded string
+     * @return a size_t containing the number of bytes expected in the decoded string.
+    */
+    size_t
+    calculate_decode_length(const std::string& base_64_input)
+    {
```
Comment (ebruck) [inline utils/crypto.cpp]: What if input is empty?

File: utils/crypto.cpp [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        auto padding = (base_64_input[len - 1] == '=' && base_64_input[len - 2] == '=') ? 2 : 1;
+        return (len * 3) / 4 - padding;
+    }
+
+
+    void
+    compose_and_log_OpenSSL_error(const std::string& failing_function_name)
+    {
+        char buffer[126]{0};
+        ERR_error_string_n(ERR_get_error(), buffer, 126);
```
Comment (ebruck) [inline utils/crypto.cpp]: I think you can do sizeof(buffer)

File: utils/crypto.cpp [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+     * Given the public key string in PEM format, this function will return the
+     * RSA public key structure that OpenSSL functions use.
+     *
+     * @param key a const std::string reference containing the public key in PEM format
+     * @return a pointer to the RSA structure contqining the public key
+     */
+    RSA* create_public_RSA(const std::string& key)
+    {
+        std::unique_ptr<BIO,std::function<void(BIO*)>> keybio(BIO_new_mem_buf((void*)(key.c_str()), -1), BIO_free);
+        if (nullptr==keybio)
```
Comment (ebruck) [inline utils/crypto.cpp]: if (keybio) is all you need.

File: utils/crypto.cpp [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    {
+        // TODO retrieve the public key from the contract
+        return temporary_public_pem;
+    }
+
+
+    int
+    base64_decode(const std::string& base64_message, std::vector<unsigned char>& decoded_message)
+    {
+        const auto decoded_length = calculate_decode_length(base64_message);
```
Comment (ebruck) [inline utils/crypto.cpp]: this isn't the kind of places auto is used for

File: utils/crypto.cpp [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        // it probably doesn't.
+        std::string nonconst{base64_message};
+        BIO* bio = BIO_new_mem_buf(nonconst.data(), -1);
+
+        BIO* b64 = BIO_new(BIO_f_base64());
+        bio = BIO_push(b64, bio);
+
+        BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL); //Do not use newlines to flush buffer
+
+        const auto length = BIO_read(bio, decoded_message.data(), static_cast<int>(base64_message.size()));
```
Comment (ebruck) [inline utils/crypto.cpp]: same here

File: README.md (hunk 1 above)
Comment (rnistuk) [reply README.md]: Yes. Thanks for the suggestion!

Comment (rnistuk) [reply README.md]: OK, did that refactoring and pushed.

File: bootstrap/bootstrap_peers.cpp (hunk 2 above)
Comment (rnistuk) [reply bootstrap/bootstrap_peers.cpp]: Right! OK, as penance I have scoured the code base and corrected this everywhere I found it.

File: raft/raft.cpp (hunk 3 above)
Comment (rnistuk) [reply raft/raft.cpp]: Fixed. ibid.

File: utils/crypto.cpp (hunk 5 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Added a test to the first line of the function that returns 0 if base_64_input is empty.

Also went back up the stack to the base64_decode function to return a failure and throw up a log message if the input that is supposed to contain a base64 encoded string is empty.

Added a unit test as well.

File: utils/crypto.cpp (hunk 6 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Yes, I can. Done.

File: utils/crypto.cpp (hunk 8 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Right, I'll stop using auto for pod.

File: utils/crypto.cpp (hunk 9 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Changed auto to int

**Pull request merged**